pkg/containers/images.go

```diff
@@ -137,42 +148,68 @@ func (c *Client) Find(ctx context.Context) (Results, error) {
 		image.parseTags()
 		err := image.populateNewest()
 		if err != nil {
-			return Results{}, err
+			return nil, err
 		}
 	}
-	return Results{
+	return &Results{
 		Images:    images,
 		ErrImages: errored,
 	}, nil
 }
 
-// getContainerImages fetches all pods and returns a slice of container images
-func (c *Client) getContainerImages() ([]string, error) {
-	klog.V(3).Infof("Getting container images from pods")
+// topControllerGetter was extract out to facilitate mocking controller.GetAllTopControllers function for testing
+type topControllerGetter = func(ctx context.Context, dynamicClient dynamic.Interface, restMapper meta.RESTMapper, namespace string) ([]controller.Workload, error)
 
-	k := c.Kube.Client
-	pods, err := k.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+// getContainerImages fetches all pods and returns a slice of container images
+func (c *Client) getContainerImages(topControllerGetter topControllerGetter) (map[string][]Workload, error) {
+	klog.V(3).Infof("Getting all top controllers from cluster")
+	topControllers, err := topControllerGetter(context.TODO(), c.Kube.DynamicClient, c.Kube.RESTMapper, "")
 	if err != nil {
-		return nil, errors.Wrap(err, "getting all pods")
+		return nil, err
 	}
-
-	imagesFound := make([]string, 0)
-	for _, pod := range pods.Items {
-		if len(pod.Spec.InitContainers) > 0 {
-			for _, container := range pod.Spec.InitContainers {
-				if container.Image != "" {
-					imagesFound = append(imagesFound, container.Image)
+	images := make(map[string][]Workload, 0)
+	for _, w := range topControllers {
+		for _, unstructuredPod := range w.Pods {
+			pod, err := toV1Pod(unstructuredPod)
+			if err != nil {
+				return nil, fmt.Errorf("unable to parse Pod from unstructured object: %w", err)
+			}
+			if len(pod.Spec.InitContainers) > 0 {
+				for _, container := range pod.Spec.InitContainers {
+					if container.Image != "" {
+						images[container.Image] = append(images[container.Image], Workload{
+							Name:      w.TopController.GetName(),
+							Namespace: w.TopController.GetNamespace(),
+							Kind:      w.TopController.GetKind(),
+							Container: container.Name,
+						})
+					}
 				}
 			}
-		}
-		for _, container := range pod.Spec.Containers {
-			if container.Image != "" {
-				imagesFound = append(imagesFound, container.Image)
+			for _, container := range pod.Spec.Containers {
+				if container.Image != "" {
+					images[container.Image] = append(images[container.Image], Workload{
+						Name:      w.TopController.GetName(),
+						Namespace: w.TopController.GetNamespace(),
+						Kind:      w.TopController.GetKind(),
+						Container: container.Name,
+					})
+				}
 			}
+			break // just need to check the first pod (to avoid workload duplication)
```

Comment:
I'd suggest explicitly looking at `Pods[0]` rather than doing a `break` for clarity.

Reply:
done